### PR TITLE
PullService lock via pullID

### PR DESCRIPTION
**Pull request by 6543**
~~part of #9307 ... but worth own pull~~ now independent

### CODE REVIEW

File: services/pull/pull.go [hunk 1]
```diff
@@ -25,9 +25,13 @@ import (
 	"code.gitea.io/gitea/modules/notification"
 	"code.gitea.io/gitea/modules/process"
 	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/sync"
 	issue_service "code.gitea.io/gitea/services/issue"
 )
 
+// TODO: use clustered lock (unique queue? or *abuse* cache)
```
Comment (6543) [inline services/pull/pull.go:32]: ? cc @lunny @zeripath ...

Comment (lunny) [reply services/pull/pull.go:32]: We haven't introduce a distributed transaction in Gitea. But we really need one.
unique queue is a temporarily method to handle that I think.

File: services/pull/check.go [hunk 2]
```diff
@@ -314,6 +314,8 @@ func handle(data ...queue.Data) []queue.Data {
 }
 
 func testPR(id int64) {
+	pullWorkingPool.CheckIn(fmt.Sprint(id))
```
Comment (lunny) [inline services/pull/check.go:320]: We should really reduce to use this memory pool mechanism.

**Review by delvh — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

File: services/pull/check.go (hunk 2 above)
Comment (6543) [reply services/pull/check.go:320]: we already have it ... I'll create a dedicated issue and will work on replace the whole sync package ...

**Comment by 6543** [comment]
> > We should really reduce to use this memory pool mechanism.

> we already have it ... I'll create a dedicated issue and will work on replace the whole sync package ...

**Comment by zeripath** [comment]
hmm... Isn't the correct place to lock this kind of thing the db?

**Comment by 6543** [comment]
Please comment on  #19620

**Pull request merged**